modules/tinymce/src/themes/silver/main/ts/ui/core/complex/utils/FormatDetection.ts

```diff
@@ -7,22 +7,16 @@
 
 import { Arr, Optional } from '@ephox/katamari';
 import Editor from 'tinymce/core/api/Editor';
+import { BasicSelectItem } from '../SelectDatasets';
 
-const findNearest = (editor: Editor, getStyles, parents: Element[]) => {
+export const findNearest = (editor: Editor, getStyles: () => BasicSelectItem[]) => {
   const styles = getStyles();
+  const formats = Arr.map(styles, (style) => style.format);
 
-  return Arr.findMap(parents, (parent) => Arr.find(styles, (fmt) => editor.formatter.matchNode(parent, fmt.format))).orThunk(() => {
+  return Optional.from(editor.formatter.closest(formats)).bind((fmt) =>
+    Arr.find(styles, (data) => data.format === fmt)
+  ).orThunk(() => {
     if (editor.formatter.match('p')) { return Optional.some({ title: 'Paragraph', format: 'p' }); }
```

Comment:
Not sure if you wanted to clean this up while your here, as it's unrelated to your changes (so feel free to ignore), but this could be simplified to `Optionals.someIf(editor.formatter.match('p'), { title: 'Paragraph', format: 'p' })`